core/node/vm_runner/src/output_handler.rs

```diff
@@ -0,0 +1,541 @@
+                sender,
+            }) => {
+                self.internal = Some(OutputHandlerState::Finished);
+                sender
+                    .send(Box::pin(async move {
+                        handler.handle_l1_batch(updates_manager).await
+                    }))
+                    .ok();
+                Ok(())
+            }
+            Some(OutputHandlerState::Finished) => {
+                self.internal = state;
+                Err(anyhow::anyhow!("Cannot handle any more L1 batches"))
+            }
+            None => Err(anyhow::anyhow!(
+                "Unexpected state, missing output handler state"
+            )),
+        }
+    }
+}
+
+/// A runnable task that continually awaits for the very next unprocessed batch to be processed and
+/// marks it as so using `Io`.
+pub struct ConcurrentOutputHandlerFactoryTask<Io: VmRunnerIo> {
+    pool: ConnectionPool<Core>,
+    io: Io,
+    state: Arc<DashMap<L1BatchNumber, oneshot::Receiver<BoxFuture<'static, anyhow::Result<()>>>>>,
+}
+
+impl<Io: VmRunnerIo> Debug for ConcurrentOutputHandlerFactoryTask<Io> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ConcurrentOutputHandlerFactoryTask")
+            .field("pool", &self.pool)
+            .field("io", &self.io)
+            .finish()
+    }
+}
+
+impl<Io: VmRunnerIo> ConcurrentOutputHandlerFactoryTask<Io> {
+    /// Starts running the task which is supposed to last until the end of the node's lifetime.
+    ///
+    /// # Errors
+    ///
+    /// Propagates DB errors.
+    pub async fn run(self, stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        const SLEEP_INTERVAL: Duration = Duration::from_millis(50);
+
+        let mut conn = self.pool.connection_tagged(Io::name()).await?;
+        let mut latest_processed_batch = self.io.latest_processed_batch(&mut conn).await?;
+        drop(conn);
+        loop {
+            if *stop_receiver.borrow() {
+                tracing::info!("`ConcurrentOutputHandlerFactoryTask` was interrupted");
+                return Ok(());
+            }
+            match self.state.remove(&(latest_processed_batch + 1)) {
+                None => {
+                    tracing::debug!(
+                        "Output handler for batch #{} has not been created yet",
+                        latest_processed_batch + 1
+                    );
+                    tokio::time::sleep(SLEEP_INTERVAL).await;
+                }
+                Some((_, receiver)) => {
+                    // Wait until the future is sent through the receiver, happens when
+                    // `handle_l1_batch` is called on the corresponding output handler
+                    let future = receiver.await?;
+                    // Wait until the future is completed, meaning that the `handle_l1_batch`
+                    // computation has finished, and we can consider this batch to be completed
+                    future.await?;
```

Comment:
If I understand correctly, this will make `handle_l1_batch()` not concurrent; futures don't do work unless polled, and you poll these futures one-by-one here. Is this intentional? To achieve concurrent processing, one way is spawning Tokio tasks and using their handles instead of futures; tasks progress independently of whether their handles are polled.

Reply:
🤦 

I had a feeling the tests were not running concurrently, thanks for the great catch